ChopChop/ViewModels/RecipeFormViewModel.swift

```diff
@@ -0,0 +1,214 @@
+                guard let self = self else {
+                    return
+                }
+                switch value {
+                case .failure:
+                    self.allRecipeCategories = []
+                case .finished:
+                    break
+                }
+            },
+            receiveValue: { [weak self] categories in
+                guard let self = self else {
+                    return
+                }
+                self.allRecipeCategories = categories
+            })
+            .store(in: &recipeCategoryCancellable)
+    }
+
+    func ensureValidServing() {
+        let filtered = serving.filter { "0123456789.".contains($0) }
+        if filtered != serving {
+            serving = filtered
+        }
+    }
+
+    func checkFormValid() throws {
+        if recipeName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+            throw RecipeFormError.emptyName
+        }
+
+        if serving.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+            throw RecipeFormError.emptyServing
+        }
+
+        if steps.isEmpty {
+            throw RecipeFormError.emptyStep
+        }
+
+        for step in steps {
+            if step.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+                throw RecipeFormError.emptyStepDescription
+            }
+        }
+
+        if ingredients.isEmpty {
+            throw RecipeFormError.emptyIngredient
+        }
+
+        for ingredient in ingredients {
+            if ingredient.amount.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+                throw RecipeFormError.emptyIngredientQuantity
+            }
+
+            if ingredient.ingredientName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+                throw RecipeFormError.emptyIngredientDescription
+            }
+        }
+    }
+
+    func parseData() {
+        let parsedIngredients = RecipeParser.parseIngredientString(ingredientString: ingredientParsingString)
+            .map({
+                RecipeIngredientRowViewModel(
+                    amount: $0.value.value.description,
+                    unit: $0.value.type,
+                    ingredientName: $0.key
+                )
+            })
+        let parsedSteps = RecipeParser.parseInstructions(instructions: instructionParsingString)
+        ingredients.append(contentsOf: parsedIngredients)
+        steps.append(contentsOf: parsedSteps)
+        instructionParsingString = ""
+        ingredientParsingString = ""
+    }
+
+    func saveRecipe() -> Bool {
+        do {
+            try checkFormValid()
+            var newRecipe = try generateRecipe()
+            try storageManager.saveRecipe(&newRecipe)
+            return true
+        } catch {
+            hasError = true
+            setErrorMessage(error: error)
+            return false
+        }
+    }
+
+    private func setErrorMessage(error: Error) {
```

Comment:
not v sure how this part works, but i think we should display all errors at once